File: sdkwork-rtc-sdk-python/sdkwork_rtc_sdk/driver_manager.py
from .provider_activation_catalog import get_rtc_provider_activation_by_provider_key
from .provider_catalog import (
    DEFAULT_RTC_PROVIDER_KEY,
    RtcProviderCatalog,
    get_rtc_provider_by_provider_key,
)
from .provider_selection import (
    RtcProviderSelection,
    RtcProviderSelectionRequest,
    resolve_rtc_provider_selection,
)
from .provider_support import (
    RtcProviderSupport,
    RtcProviderSupportStateRequest,
    create_rtc_provider_support_state,
)


class RtcDriverManager:
    def resolveSelection(
        self,
        request: RtcProviderSelectionRequest | None = None,
        *,
        defaultProviderKey: str = DEFAULT_RTC_PROVIDER_KEY,
    ) -> RtcProviderSelection:
        return resolve_rtc_provider_selection(
            request,
            default_provider_key=defaultProviderKey,
        )

    def describeProviderSupport(self, providerKey: str) -> RtcProviderSupport:
        official = get_rtc_provider_by_provider_key(providerKey) is not None
        activation = get_rtc_provider_activation_by_provider_key(providerKey)

        return create_rtc_provider_support_state(
            RtcProviderSupportStateRequest(
                providerKey=providerKey,
                builtin=activation.builtin if activation is not None else False,
                official=official,
                registered=activation.runtimeBridge if activation is not None else False,
            )
        )

    def listProviderSupport(self) -> list[RtcProviderSupport]:
        return [
            self.describeProviderSupport(entry.providerKey)
            for entry in RtcProviderCatalog.entries
        ]

File: sdkwork-rtc-sdk-rust/src/driver_manager.rs
use crate::provider_activation_catalog::get_rtc_provider_activation_by_provider_key;
use crate::provider_catalog::{get_rtc_provider_by_provider_key, OFFICIAL_RTC_PROVIDERS};
use crate::provider_selection::{
    resolve_rtc_provider_selection, RtcProviderSelection, RtcProviderSelectionRequest,
};
use crate::provider_support::{
    create_rtc_provider_support_state, RtcProviderSupport, RtcProviderSupportStateRequest,
};

pub struct RtcDriverManager;

impl RtcDriverManager {
    #[allow(non_snake_case)]
    pub fn resolveSelection(
        &self,
        request: &RtcProviderSelectionRequest,
        defaultProviderKey: Option<&str>,
    ) -> RtcProviderSelection {
        resolve_rtc_provider_selection(request, defaultProviderKey)
    }

    #[allow(non_snake_case)]
    pub fn describeProviderSupport(&self, providerKey: &str) -> RtcProviderSupport {
        let official = get_rtc_provider_by_provider_key(providerKey).is_some();
        let activation = get_rtc_provider_activation_by_provider_key(providerKey);

        create_rtc_provider_support_state(RtcProviderSupportStateRequest {
            providerKey: providerKey.to_string(),
            builtin: activation.is_some_and(|entry| entry.builtin),
            official,
            registered: activation.is_some_and(|entry| entry.runtimeBridge),
        })
    }

    #[allow(non_snake_case)]
    pub fn listProviderSupport(&self) -> Vec<RtcProviderSupport> {
        OFFICIAL_RTC_PROVIDERS
            .iter()
            .map(|entry| self.describeProviderSupport(entry.providerKey))
            .collect()
    }
}
